From 449d95f213cb525bdad81480bf74b7586d4ecaa7 Mon Sep 17 00:00:00 2001
From: scateu <scateu@gmail.com>
Date: Sat, 18 Jul 2020 17:51:17 +0800
Subject: [PATCH] disable double fork to make ssh work in macOS

See: 

https://groups.google.com/forum/m/#!topic/comp.mail.pine/PZQ5ndvyzjE
https://comp.mail.pine.narkive.com/yy2fdOns/can-t-get-ssh-to-work-with-imap
http://graphics.stanford.edu/lab/mail/unixclients.html#tunnel

The minimum working config of alpine is:                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                
Rsh Command                         = <No Value Set> 
Rsh Path                            = <No Value Set> 
Rsh Open Timeout                    = 0 
Ssh Command                         = <No Value Set> 
Ssh Path                            = /usr/bin/ssh 
Ssh Open Timeout                    = <No Value Set>                                    

On remote machine: 

$ ln -s /path/to/imapd /etc/rimapd

---
 imap/src/osdep/unix/tcp_unix.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/imap/src/osdep/unix/tcp_unix.c b/imap/src/osdep/unix/tcp_unix.c
index bc5fe33..2234ead 100644
--- a/imap/src/osdep/unix/tcp_unix.c
+++ b/imap/src/osdep/unix/tcp_unix.c
@@ -413,8 +413,8 @@ TCPSTREAM *tcp_aopen (NETMBX *mb,char *service,char *usrbuf)
     return NIL;
   }
   if (!i) {			/* if child */
-    alarm (0);			/* never have alarms in children */
-    if (!fork ()) {		/* make grandchild so it's inherited by init */
+    //alarm (0);			/* never have alarms in children */
+    //if (!fork ()) {		/* make grandchild so it's inherited by init */
       int cf;			/* don't alter parent vars in case vfork() */
       int maxfd = max (20,max (max(pipei[0],pipei[1]),max(pipeo[0],pipeo[1])));
       dup2 (pipei[1],1);	/* parent's input is my output */
@@ -424,12 +424,12 @@ TCPSTREAM *tcp_aopen (NETMBX *mb,char *service,char *usrbuf)
       for (cf = 3; cf <= maxfd; cf++) close (cf);
       setpgrp (0,getpid ());	/* be our own process group */
       _exit (execv (path,argv));/* now run it */
-    }
-    _exit (1);			/* child is done */
+    //}
+    //_exit (1);			/* child is done */
   }
-  grim_pid_reap (i,NIL);	/* reap child; grandchild now owned by init */
-  close (pipei[1]);		/* close child's side of the pipes */
-  close (pipeo[0]);
+  //grim_pid_reap (i,NIL);	/* reap child; grandchild now owned by init */
+  //close (pipei[1]);		/* close child's side of the pipes */
+  //close (pipeo[0]);
 
 				/* create TCP/IP stream */
   stream = (TCPSTREAM *) memset (fs_get (sizeof (TCPSTREAM)),0,
-- 
2.27.0

